content/blog/2022-11-hackathon-2022-group-chat.md

```diff
@@ -0,0 +1,208 @@
+
+```javascript
+import getApiClient from '@wazo/sdk/lib/service/getApiClient';
+
+const client = getApiClient('hostname.example');
+
+// On page load, we fetch rooms for the current user
+const rooms = await client.chatd.getUserRooms();
+
+// On room load, we fetch previous message in a room
+const messages = await client.chatd.getRoomMessages(ROOM_UUID);
+
+// On form submit, we send the new message to chatd services
+await client.chatd.sendRoomMessage(ROOM_UUID, {
+  content: 'The new message content',
+  alias: 'Alias/Name of the member',
+});
+```
+
+The second part is about listening to WebSocket messages. We used the wazo-js-sdk WebSocket client.
+Here are the events that we listened on :
+
+```javascript
+// Initialize websocket client
+import { WazoWebSocketClient } from '@wazo/sdk';
+
+const ws = new WazoWebSocketClient({
+  host,
+  token: response.token,
+  events: [
+    'chatd_user_room_message_created',
+    // Other events to listen
+  ],
+  version: '2',
+});
+
+// Main events to listen
+ws.on('chatd_user_room_message_created', (message) => {
+  // Listen to "room message created" events
+  // - Validate the message is in the displayed room
+  // - Add message to all messages
+  //
+  // Message Payload Example:
+  // {
+  //   room_uuid: "697a35a6-534c-461d-9466-6f77d0181e80",
+  //   data: {
+  //     content: 'The new message content',
+  //     alias: 'Alias/Name of the member',
+  //     created_at: '2022-10-24T15:50:00.000000+00:00',
+  //     user_uuid: '8040ec9d-1a61-4ca3-abe5-ad7c2f192e03',
+  //     uuid: 'a0e7dc92-92a3-485b-b8dd-09a909a1f5a0',
+  //   }
+  // }
+});
+
+ws.on('chatd_user_room_created', (message) => {
+  // Listen to "room created" events
+  // - Add new room to the listing
+  //
+  // Message Payload Example
+  // {
+  //   uuid: '6f9c7df4-8529-4336-9b65-21ceec774583'
+  //   name: 'New Room Name',
+  //   tenant_uuid: '47e1a830-cb3a-4ac9-9117-82acbb54d72b',
+  //   users: [
+  //     {
+  //       tenant_uuid: '5d5e7e18-46ae-4046-beda-3cdcd5449142',
+  //       uuid: '551d06a1-808a-417d-995d-7abb66d55c33',
+  //     },
+  //     // Other users
+  //   ]
+  // }
+});
+```
+
+### The Hardest Parts of a Group Chat Application
+
+We found that the most challenging parts of a group chat application are :
+
+- Creating a feature-rich interface. We maybe don’t realize it at first, but chat apps are powerful
+  tools with a lot of shortcuts. Shortcut examples: editing previous message, switch to new
+  channels, alias in message (@mention, emoji syntax `:smile:`, link detection, etc.)
+- Other details not regarding the core feature (send and receiving messages). Examples: threads,
+  notifications, participants management, privacy, etc.
+
+## Conclusion
+
+The hackathon was a great opportunity for us to work closely with our colleagues and try new
+technologies. Also, it gives us good feedback on what is missing and how much efforts we should
+deploy to implement a complete group chat application.
```

Comment:
It also fives us a good idea about what is missing and how much effort would be required to implement and deploy these feratures